colcon_defaults/argument_parser/defaults.py

```diff
@@ -25,6 +26,23 @@
     'Set path to the yaml file containing the default values for the command '
     'line arguments (default: $COLCON_HOME/defaults.yaml)')
 
+"""File with default values specific to the workspace."""
+WORKSPACE_DEFAULTS_FILE = 'colcon_defaults.yaml'
+
+
+def _deep_update(source, overrides):
+    """
+    Update a nested dictionary or similar mapping.
+
+    Modifies ``source`` in place with values from ``overrides``. Taken from
+    https://stackoverflow.com/a/30655448
+    """
+    for key, value in overrides.items():
+        if isinstance(value, collections.abc.Mapping) and value:
+            _deep_update(source.get(key, {}), value)
```

Comment:
This isn't working as intended if `source` doesn't already contain the mapping. `source.get(key, {})` will give the recursive call a mapping to add values to, but then it isn't getting put back into the original `source` mapping.

```suggestion
            source.setdefault(key, {})
            _deep_update(source[key], value)
```

Without a fix here, this change doesn't work on my local machine, where I have no global colcon `defaults.yaml` to begin with.

Reply:
Glad you were able to try and find this too. Looks like my local testing always keep the global file and I never tried removing it completely. Your elegant suggested solution is included now and I confirmed it on my machine as well.

Reply:
Revisiting that stackoverflow question where this originally came from it's insane the number of variations people proposed to "correct" subtle issues with the accepted answer, most of which probably contain their own subtle issues.